### ISSUE TITLE
SignalR Retry lost connection behavior change

### ISSUE THREAD

**Issue by hridpath** [dotnet/aspnetcore#45810]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am trying to use the RetryPolicy to activate when an app starts but the server is not available. Currently the only time a retry is done is after a connection has been established and then gets disconnected.

### Describe the solution you'd like

It would be nice to have a property on the RetryPolicy to allow for use in the StartAsyncCore even when there has not been a connection established.

So if I have a AutoReconnect(RetryPolicy) in the hub builder, and the first attemp to connect to the server fails, We would like the Retry Policy to activate.

**Comment by hridpath** [comment]
### Here are examples of what could be done:
So imagine a factory with may stations(Clients) and one Server. In the scenario where the power is off, for whatever reason, the stations can boot into the application(Client) quickly. The Server takes maybe a couple of minutes to get all the services ready. So the Server Hub  is not immediately available. In the current code. The clients would require a User input to restart or code added to  loop and wait for the connection. If the Client HubConnection has a RetryPolicy, we should not have to write extra code to handle this scenario.
**So a boolean property can be added to the IRetryPolicy:**
` /// <summary>
    /// When true this property tells the HubConnetion to run this RetryPolicy when an initial connection cannot be established. This allows for
    /// clients to attempt to connect to a server that may be late in starting up, or failed on startup and requires a restart. The clients can then
    /// reconnect when the server becomes available, if NextRetryDelay has not receives a null.
    /// </summary>`
    **bool? RetryOnStart { get; set; } => false;**

**Then by making three changes to the HubConnection code:**
Change this:
**`private async Task StartAsyncCore(CancellationToken cancellationToken)`**
To this:
`private async Task StartAsyncCore(CancellationToken cancellationToken,`**_bool fromReconnectAsync = false_**`)`

**Add this to the StartAsyncCore method when HandshakeAsync fails**:
` //[DS]If Handshake fails, and this is the first attempt, start ReconnectAsync only once.
  //[CAUTION] The ReconnectAsync method calls this method so there is a chase scenario.`
            **_if (_reconnectPolicy.RetryOnStart && !fromReconnectAsync)
                await ReconnectAsync(ex);_**
**Then lastly, change the call arguments to StartAsyncCore in the ReconnectAsync method from:**
**_await StartAsyncCore(_state.StopCts.Token).ConfigureAwait(false);_**
to:
**_await StartAsyncCore(_state.StopCts.Token, true).ConfigureAwait(false);_**

**Comment by BrennanConroy** [comment]
Today we recommend writing your own start loop if this is the behavior you desire. https://learn.microsoft.com/aspnet/core/signalr/dotnet-client?view=aspnetcore-7.0&tabs=visual-studio#automatically-reconnect

This is because it's a lot easier to reason that if someone already had an active connection they would want it to reconnect automatically.
But if someone hasn't connected yet there could be multiple reasons that reconnect wouldn't be wanted, e.g. wrong url, not authenticated, not connected to internet. If we did something like this, we'd likely go the opt-in route like you mentioned with `IRetryPolicy.RetryOnStart`.

We'll backlog and see if there is any more community interest in this behavior.
